File: floyd.py
# floyd.py
# Algoritmo de Floyd-Warshall usando NetworkX

import networkx as nx

class Floyd:
    def __init__(self, grafo):
        self.grafo = grafo
        self.pred  = {}
        self.dist  = {}
        self.ejecutar()

    def ejecutar(self):
        self.pred, self.dist = nx.floyd_warshall_predecessor_and_distance(
            self.grafo.G, weight="weight"
        )

    def get_distancia(self, origen, destino):
        try:
            d = self.dist[origen][destino]
            return None if d == float("inf") else int(d)
        except KeyError:
            return None

    def get_ruta(self, origen, destino):
        try:
            nodos = nx.reconstruct_path(origen, destino, self.pred)
            return " -> ".join(nodos)
        except (KeyError, nx.NetworkXError):
            return f"No existe ruta de {origen} a {destino}."

    def get_excentricidad(self, ciudad):
        nodos = self.grafo.listar_ciudades()
        max_dist = 0
        for origen in nodos:
            if origen == ciudad:
                continue
            d = self.dist.get(origen, {}).get(ciudad, float("inf"))
            if d == float("inf"):
                return float("inf")
            if d > max_dist:
                max_dist = d
        return int(max_dist)

    def get_centro(self):
        nodos = self.grafo.listar_ciudades()
        min_ecc = float("inf")
        centro  = None
        for ciudad in nodos:
            ecc = self.get_excentricidad(ciudad)
            if ecc < min_ecc:
                min_ecc = ecc
                centro  = ciudad
        return centro, min_ecc

    def imprimir_apsp(self):
        nodos = self.grafo.listar_ciudades()
        ancho = max((len(n) for n in nodos), default=5) + 2
        col   = 10
        print("\n===== MATRIZ APSP (Caminos mas cortos) =====")
        print(f"{'':>{ancho}}", end="")
        for n in nodos:
            print(f"{n:>{col}}", end="")
        print()
        for origen in nodos:
            print(f"{origen:>{ancho}}", end="")
            for destino in nodos:
                d = self.dist.get(origen, {}).get(destino, float("inf"))
                print(f"{'INF' if d == float('inf') else int(d):>{col}}", end="")
            print()
        # Fila de excentricidades
        print(f"{'Excentricidad':>{ancho}}", end="")
        for ciudad in nodos:
            ecc = self.get_excentricidad(ciudad)
            print(f"{'INF' if ecc == float('inf') else ecc:>{col}}", end="")
        print()
        print("============================================\n")

File: Floyd.java
public class Floyd {

    private int[][] dist;
    private int[][] next;
    private int     n;
    private Grafo   grafo;

    public Floyd(Grafo grafo) {
        this.grafo = grafo;
        ejecutar();
    }

    public void ejecutar() {
        n = grafo.getNumVertices();
        int[][] C = grafo.getMatriz();

        dist = new int[n][n];
        next = new int[n][n];

        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++) {
                dist[i][j] = C[i][j];
                next[i][j] = (i != j && C[i][j] < Grafo.INF) ? j : -1;
            }

        for (int k = 0; k < n; k++)
            for (int i = 0; i < n; i++)
                for (int j = 0; j < n; j++)
                    if (dist[i][k] < Grafo.INF && dist[k][j] < Grafo.INF
                            && dist[i][k] + dist[k][j] < dist[i][j]) {
                        dist[i][j] = dist[i][k] + dist[k][j];
                        next[i][j] = next[i][k];
                    }
    }

    public int getDistancia(String origen, String destino) {
        int i = grafo.getIndice(origen);
        int j = grafo.getIndice(destino);
        return (i < 0 || j < 0) ? Grafo.INF : dist[i][j];
    }

    public String getRuta(String origen, String destino) {
        int i = grafo.getIndice(origen);
        int j = grafo.getIndice(destino);
        if (i < 0 || j < 0)           return "Ciudad no encontrada.";
        if (i == j)                    return origen;
        if (dist[i][j] >= Grafo.INF)  return "No existe ruta de " + origen + " a " + destino + ".";

        StringBuilder sb = new StringBuilder(grafo.getCiudad(i));
        int actual = i;
        while (actual != j) {
            actual = next[actual][j];
            if (actual < 0) return "Error al reconstruir ruta.";
            sb.append(" -> ").append(grafo.getCiudad(actual));
        }
        return sb.toString();
    }

    public int getExcentricidad(String ciudad) {
        int col = grafo.getIndice(ciudad);
        if (col < 0) return Grafo.INF;
        int max = 0;
        for (int fila = 0; fila < n; fila++) {
            if (fila == col) continue;
            if (dist[fila][col] >= Grafo.INF) return Grafo.INF;
            if (dist[fila][col] > max) max = dist[fila][col];
        }
        return max;
    }

    public String getCentro() {
        int    minEcc = Grafo.INF;
        String centro = null;
        for (int i = 0; i < n; i++) {
            int ecc = getExcentricidad(grafo.getCiudad(i));
            if (ecc < minEcc) { minEcc = ecc; centro = grafo.getCiudad(i); }
        }
        return centro;
    }

    public void imprimirAPSP() {
        System.out.println("\n===== MATRIZ APSP (Caminos mas cortos) =====");
        System.out.printf("%-20s", "");
        for (int j = 0; j < n; j++)
            System.out.printf("%-14s", cortar(grafo.getCiudad(j), 12));
        System.out.println();
        for (int i = 0; i < n; i++) {
            System.out.printf("%-20s", cortar(grafo.getCiudad(i), 18));
            for (int j = 0; j < n; j++)
                System.out.printf("%-14s", dist[i][j] >= Grafo.INF ? "INF" : dist[i][j]);
            System.out.println();
        }
        System.out.printf("%-20s", "Excentricidad");
        for (int j = 0; j < n; j++) {
            int ecc = getExcentricidad(grafo.getCiudad(j));
            System.out.printf("%-14s", ecc >= Grafo.INF ? "INF" : ecc);
        }
        System.out.println();
        System.out.println("============================================\n");
    }

    public int[][] getDist() { return dist; }
    public int[][] getNext() { return next; }

    private String cortar(String s, int max) {
        return s.length() > max ? s.substring(0, max - 2) + ".." : s;
    }
}
